### PR TITLE
fix: infer oop form for SDEProblem/SDEFunction with StaticArrays

### LINKED ISSUE #2814: Unable to simualte SDEs with static array u0/ps
This is important since this is required for GPU simulations.

MWE:
```julia
using ModelingToolkit, StaticArrays, StochasticDiffEq

# Create SDE system.
@parameters p d 
@variables t X(t)
D = Differential(t)
eqs = [D(X) ~ p - d*X]
noise_eqs = [sqrt(p), sqrt(p - d*X)]
@named ssys = SDESystem(eqs, noise_eqs, t, [X], [p, d])
ssys = complete(ssys)

# Create SDEProblem
u0 = @SVector [X => 10.0f0]
tspan = (0.0f0, 10.0f0)
ps = @SVector [p => 2.0f0, d => 0.5f0]
sprob = SDEProblem(ssys, u0, tspan, ps)
solve(sprob, ImplicitEM())
```
Gives an `ERROR: Initial condition incompatible with functional form.` error.

For ODEs, this work:
```julia
using OrdinaryDiffEq
@named osys = ODESystem(eqs, t, [X], [p, d])
osys = complete(osys)

# Create SDEProblem
u0 = @SVector [X => 10.0f0]
tspan = (0.0f0, 10.0f0)
ps = @SVector [p => 2.0f0, d => 0.5f0]
oprob = ODEProblem(osys, u0, tspan, ps)
solve(oprob, Tsit5())
```

### PR DESCRIPTION
**Pull request by AayushSabharwal**
Close #2814 

## Checklist

- [ ] Appropriate tests were added
- [ ] Any code changes were done in a way that does not break public API
- [ ] All documentation related to code changes were updated
- [ ] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [ ] Any new documentation only uses public API

### CODE REVIEW

**Comment by AayushSabharwal** [comment]
The problem here is that [this](https://github.com/SciML/StochasticDiffEq.jl/blob/4af4ab8382c257bf2b802aafca4f15baa20cf08b/src/perform_step/sdirk.jl#L30) line doesn't work with static arrays. `K` is an `SVector{1}` whereas `L` is an `SMatrix{1, 1}` (since noise is a matrix). I guess that line should have an `@..`?

```julia
julia> [1;;] + [1]
1×1 Matrix{Int64}:
 2
julia> @SArray[1;;] + @SArray[1]
ERROR: DimensionMismatch: Sizes (Size(1, 1), Size(1,)) of input arrays do not match
```

**Comment by ChrisRackauckas** [comment]
That's a method only for diagonal noise, so that method won't work on array

**Comment by AayushSabharwal** [comment]
1x1 noise is diagonal, right? Which I guess is why I guess it hits that

**Comment by ChrisRackauckas** [comment]
Yeah, what we should do is detect that `if isdiagonal` then turn it it into a vector so it represents diagonal noise.

**Review by ChrisRackauckas — APPROVED** [review]

**Comment by TorkelE** [comment]
This update causes Catalyst CI to fail. Given that Catalyst has a very extensive test set, and often has catched things that the MTK tests have missed, would it make sense to not merge things which cause Catalyst downstream errors?

**Comment by TorkelE** [comment]
(I have modified Catalyst so that it works with this new version. However, generally, I do think we should try and not ignore the CI that we got)

**Comment by ChrisRackauckas** [comment]
I don't think the Catalyst tests are right here?

**Comment by TorkelE** [comment]
The Catalyst tests are fine. Basically, this PR goes from requiring an internal matrix to be `Matrix{Num}` from `Matrix{Any}` (I know Sam has argued for the latter representation being used, personally I have no strong opinion). Anyway, this breaks a couple of things (which I Catalyst I switched around so that it should be fine).

**Comment by ChrisRackauckas** [comment]
Oh I thought the Catalyst issue was the sizing thing we were talking about? Requiring Matrix{Num} isn't intentional, though it is somewhat bettter.

**Comment by TorkelE** [comment]
No, the sizing got fixed (thanks for that!). Practically, I'm not sure of changing to `Matrix{Num}` is a big change, but next time it would be useful to at least have a note that there has been an update which breaks the Catalyst tests 👍

**Comment by isaacsas** [comment]
Shouldn’t one be able to generate SDESystems without using Nums? If not, wouldn’t this require all compositional modeling tools to rewrap internal unwrapped expressions when generating SDESystems from other systems?

File: src/systems/diffeqs/sdesystem.jl [hunk 1]
```diff
@@ -231,6 +231,9 @@ function generate_diffusion_function(sys::SDESystem, dvs = unknowns(sys),
     if isdde
         eqs = delay_to_function(sys, eqs)
     end
+    if eqs isa AbstractMatrix && isdiag(eqs)
+        eqs = diag(eqs)
+    end
```
Comment (isaacsas) [inline src/systems/diffeqs/sdesystem.jl:236]: @ChrisRackauckas @AayushSabharwal I don't think this is correct code for symbolics. The problem is that `isdiag` calls `iszero` which for a non-Num will create a symbolic boolean expression, i.e. 
```julia
julia> @variables A
e1-element Vector{Num}:
 A

julia> eq = 5*A^2 + 2*A + 65
65 + 2A + 5(A^2)

julia> iszero(eq)
false

julia> iszero(ModelingToolkit.unwrap(eq))
(65 + 2A + 5(A^2)) == 0
```
So this now explicitly breaks creating an `SDESystem` with unwrapped expressions.

Comment (isaacsas) [reply src/systems/diffeqs/sdesystem.jl:236]: Also, as @TorkelE mentioned Catalyst master tests should now generally pass, so if Catalyst tests fail when running downstream tests on an MTK PR that is meaningful again (unlike before we updated Catalyst to MTK 9).

Comment (ChrisRackauckas) [reply src/systems/diffeqs/sdesystem.jl:236]: The Catalyst v14 release hasn't been made yet, so it's still not testing the updated code?

Comment (ChrisRackauckas) [reply src/systems/diffeqs/sdesystem.jl:236]: > The problem is that isdiag calls iszero which for a non-Num will create a symbolic boolean expression, i.e.

Thanks, we need to have a different `isdiag` that uses `isequal`. @AayushSabharwal can you prioritize this?

Comment (ChrisRackauckas) [reply src/systems/diffeqs/sdesystem.jl:236]: @isaacsas is there an issue open on this? If so, let's assign @AayushSabharwal to it.

Comment (isaacsas) [reply src/systems/diffeqs/sdesystem.jl:236]: I opened https://github.com/SciML/ModelingToolkit.jl/issues/2868

Comment (isaacsas) [reply src/systems/diffeqs/sdesystem.jl:236]: > The Catalyst v14 release hasn't been made yet, so it's still not testing the updated code?

Also, doesn't this:

https://github.com/SciML/ModelingToolkit.jl/actions/runs/9772401615/job/26976712250#step:6:579

indicate it was tested against Catalyst master?

**Pull request merged**